Dockerfile

```diff
@@ -1,59 +1,84 @@
-
-FROM ruby:2.6.3 as release
+FROM ruby:2.6.3 AS base
 MAINTAINER dxw <rails@dxw.com>
+
 RUN apt-get update && apt-get install -qq -y \
   build-essential \
   libpq-dev \
   --fix-missing --no-install-recommends
-RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - \
-        && apt-get install -y nodejs
 
-ENV INSTALL_PATH /srv/beis-roda
-RUN mkdir -p $INSTALL_PATH
+ENV APP_HOME /app
+ENV DEPS_HOME /deps
 
-# This must be ordered before rake assets:precompile
-COPY --from=dependencies ./deps/node_modules $INSTALL_PATH/node_modules
-COPY --from=dependencies ./deps/node_modules/govuk-frontend/govuk/assets $INSTALL_PATH/app/assets
+ARG RAILS_ENV
+ENV RAILS_ENV ${RAILS_ENV:-production}
+ENV NODE_ENV ${RAILS_ENV:-production}
 
-WORKDIR $INSTALL_PATH
+# ------------------------------------------------------------------------------
+# dependancies
+# ------------------------------------------------------------------------------
+FROM base AS dependencies
 
-# set rails environment
-ARG RAILS_ENV
-ENV RAILS_ENV=${RAILS_ENV:-production}
-ENV RACK_ENV=${RAILS_ENV:-production}
+# Set up install environment
+RUN mkdir -p ${DEPS_HOME}
+WORKDIR ${DEPS_HOME}
+# End
 
-COPY Gemfile $INSTALL_PATH/Gemfile
-COPY Gemfile.lock $INSTALL_PATH/Gemfile.lock
+RUN curl -sL https://deb.nodesource.com/setup_10.x | bash - \
+        && apt-get install -y nodejs
+
+# Install Ruby dependencies
+COPY Gemfile ${DEPS_HOME}/Gemfile
+COPY Gemfile.lock ${DEPS_HOME}/Gemfile.lock
 
 RUN gem update --system 3.0.3
 RUN gem update rake 13.0.1
 RUN gem install bundler
 
-# bundle ruby gems based on the current environment, default to production
-RUN echo $RAILS_ENV
-RUN \
-  if [ "$RAILS_ENV" = "production" ]; then \
-    bundle install --jobs 4 --without development test --retry 10; \
+RUN bundle config set frozen 'true'
+
+# Configure bundler for the environment
+RUN if [ ${RAILS_ENV} = "production" ]; then \
+  bundle config set without 'development test'; \
+  elif [ ${RAILS_ENV} = "test" ]; then \
+  bundle config set without 'development'; \
   else \
-    bundle install --jobs 4 --retry 10; \
+  bundle config set without 'test'; \
   fi
 
-COPY . $INSTALL_PATH
+RUN bundle config
+RUN bundle install --retry 3 --jobs 4
+# end
+
+# Install JavaScript dependencies
+COPY package.json ${DEPS_HOME}/package.json
+COPY package-lock.json ${DEPS_HOME}/package-lock.json
+
+RUN npm set progress=false && npm config set depth 0
+RUN npm install
+#end
+
+# ------------------------------------------------------------------------------
+# web
+# ------------------------------------------------------------------------------
+FROM dependencies AS web
+
+# Set up install environment
+RUN mkdir -p ${APP_HOME}
+WORKDIR ${APP_HOME}
+# end
+
+# This must be ordered before rake assets:precompile
+COPY --from=dependencies ${DEPS_HOME}/Gemfile ${APP_HOME}/Gemfile
+COPY --from=dependencies ${DEPS_HOME}/Gemfile.lock ${APP_HOME}/Gemfile.lock
+COPY --from=dependencies ${GEM_HOME} ${GEM_HOME}
```

Comment:
I can't see that `GEM_HOME` is defined anywhere, it seems to build without it though so maybe we can remove this line?

Reply:
Interesting! I assumed GEM_HOME is set by ruby itself so the contianer has it already, I also assumed this is where the system gems are installed and that is location the bundler would use!

I am going to have to go and poke the container to see what I think about this one, it may also relate to using (or not using) the `--deployment` options (see below)

Reply:
Looking at it now, I am kind of thinking the whole thing is a bit odd!? 

GEM_HOME does seem to be set, so our container has it as `/usr/local/bundle` if that is the case is the third copy doing a copy from the dependency stage to the current stage which is pointless as it's the same location?

There also seems little point copying the Gemfile and Gemfile.lock as they come along later in line 77 anyway?